src/fork.mjs

```diff
@@ -0,0 +1,32 @@
+import AdvancedMapInitialized from 'advanced-map-initialized'
+import map from './map'
+import range from './range'
+import ensureIterable from './internal/ensure-iterable'
+
+export default function fork (iterable) {
+  const queue = []
+  const iterator = ensureIterable(iterable)[Symbol.iterator]()
+
+  function * part () {
+    let cursor = 0
+
+    while (true) {
+      while (cursor < queue.length) {
+        yield queue[cursor]
+        cursor += 1
+      }
+
+      const { value, done } = iterator.next()
+      if (done) break
+
+      queue.push(value)
```

Comment:
Iterables *should* return the same value every time `next()` is called after they are exhausted, but some may not. It is a good idea to capture the first done value explicitly to avoid bugs when dealing with user defined iterators/iterables.